groff/index.1
.TH index 1 "index" "Synrc Research Center" "N2O"
.SH NAME
index

.SH SYNOPSIS
N2O is an embeddable message protocol loop library for
WebSocket, MQTT and TCP servers. It provides basic
features, such as: process management; virtual nodes ring for
request processing; session, encoding, mq and cache services.

.SH USAGE
You can get rebar dependency and compile it:
.nf
$ mad get n2o
$ cd deps/n2o
$ mad dep com pla rep
.fi
MAY 2019 ©
\fI5HT\fR\& \fIISC\fR\& VER 6.7 6.6 6.5

.SH Endpoints
.LP
Here is a list of types of endpoints which are supported by EMQ and accessible to N2O apps:
WebSockets, MQTT, MQTT-SN, TCP, UDP, CoAP.
Normal use of N2O as a Web Framework or a Web Application Server is through WebSockets,
but for IoT and MQTT applications it could be served through UDP or SCTP protocols,
providing application level message delivery consistency.
By using MQTT as a transport we extend the supported set of endpoint protocols.

.SH MQTT
.LP
\fIEMQ\fR\& is an open-source MQTT broker implemented by Feng Lee.
N2O is EMQ compatible plugin that acts as predefined MQTT clients that form
a ring of virtual nodes, process all the protocol requests,
and answer to client topics according to classical RPC over MQ schema.
This list of modules forms a core of N2O that is common for
both MQTT and WebSocket versions:
\fIn2o\fR\& — N2O Application
: MQTT and WS — 10KB
\fIn2o_pi\fR\& — N2O Processes
— 4KB
\fIn2o_proto\fR\& — N2O Loop
: — 1KB
\fIn2o_ring\fR\& — N2O Ring
: Hash Ring — 1KB
.nf
$ mad app mqtt review
$ cd review
$ mad dep com pla rep
$ open http://127.0.0.1:8000
.fi
.LP
The list of MQTT specific modules:
\fIn2o_auth\fR\& — N2O MQTT Authentication
: MQTT auth module — 1KB
\fIn2o_mqtt\fR\& — MQTT DHT Virtual Node

.SH WebSocket
.LP
\fICOWBOY\fR\& is small, fast, and modern HTTP server implemented by Loïc Hoguin.
In this version, N2O loop is integrated as WebSocket
cowboy
handler that routes all incoming messages to N2O.
The list of WebSocket specific modules:
\fIn2o_stream\fR\& — COWBOY and XHR bridge
\fIn2o_cowboy\fR\& — COWBOY API
\fIn2o_ws\fR\& — WebSocket DHT Virtual Node
.nf
$ mad app web sample
$ cd sample
$ mad dep com pla rep
$ open https://127.0.0.1:8001/app/index.htm
.fi

.SH Protocols
\fIn2o_heart\fR\& — PING protocol
\fIn2o_ftp\fR\& — N2O File Protocol
: FTP — 4KB
\fIn2o_nitro\fR\& — N2O Nitrogen Web Framework Protocol
:  — 3KB

.SH Services
\fIn2o_session\fR\& — ETS sessions backend
\fIn2o_secret\fR\& — AES/CBC-128 encoder/decoder
\fIn2o_gproc\fR\& — GPROC MQ backend
\fIn2o_syn\fR\& — SYN MQ backend

.SH JavaScript
\fIutf8.js\fR\& — UTF-8 encoder/decoder
\fIieee754.js\fR\& — IEEE-754 encoder/decoder
\fIbert.js\fR\& — BERT encoder/decoder
\fIn2o.js\fR\& — N2O protocol looper
\fIheart.js\fR\& — HEART protocol
\fInitro.js\fR\& — NITRO protocol
\fIftp.js\fR\& — FTP protocol
\fImq.js\fR\& — MQTT client

.SH SPEC
.nf
-type n2o()       :: #bert{} | #json{} | #binary{} | #default{}.
-type cx()        :: #cx{}.
-type formatter() :: binary | json | bert | text | default | atom().
-type response()  :: { formatter(), binary() }.
-record(ok,      { data     :: term() }).
-record(error    { data     :: term() }.
-record(reply,   { msg      :: n2o(), req :: term(), ctx :: cx() } ).
-record(unknown, { msg      :: n2o(), req :: term(), ctx :: cx() } ).
-spec start(#pi{}) -> {pid(),term()} | #error{}.
-spec stop(term(),atom()) -> #pi{} | #error{}.
-spec proc(atom(),#pi{}) -> term().
-spec info(term(),term(),#cx{}) -> #reply{} | #unknown{}.
-record(pi, { name     :: term(),
table    :: atom(),
sup      :: atom(),
module   :: atom(),
state    :: term()  }).
-record(cx, { handlers  = [] :: list({atom(),atom()}),
actions   = [] :: list(tuple()),
req       = [] :: [] | term(),
module    = [] :: [] | atom(),
lang      = [] :: [] | atom(),
path      = [] :: [] | binary(),
session   = [] :: [] | binary(),
formatter = bert :: bert | json | atom(),
params    = [] :: [] | list(tuple()),
node      = [] :: [] | atom(),
client_pid= [] :: [] | term(),
state     = [] :: [] | term(),
from      = [] :: [] | binary(),
vsn       = [] :: [] | binary() }).
-type memtable()  :: atom().
-spec encode(tuple()) -> binary().
-spec decode(binary()) -> tuple().
-spec session(term(),term()) -> term().
-spec session(term()) -> term().
-spec cache(memtable(),term(),term()) -> term().
-spec cache(memtable(),term()) -> term().
.f

HISTORY.md
History
-------

* 4.5 — Basic Messaging with single topic and Echo
* 4.6 — Echo eliminated Basic Filter with req/rep topics
* 4.7 — DHT supervised node for server MQTT connections
* 4.9 — Subscribe offline clients
* 4.9.3 — Cleaned n2o_async, written manpage
* 5.10 — Cleanup
* 5.11 — Formatter API changes
* 5.11.1 — Unification of WebSocket and MQTT versions
* 6.3 — arraybuffer in JS client, n2o_pi, fixes, synrc/mqtt drop
* 6.4 — Error handling, minimalisation
* 6.5 — OTP 21.0, n2o.dev, WebSocket ring, /-endpoint for static/WS, Erlang logger



samples/apps/review/src/index.erl
-module(index).
-compile(export_all).
-include_lib("kvs/include/entry.hrl").
-include_lib("nitro/include/nitro.hrl").
-include_lib("n2o/include/wf.hrl").

main() ->
    case wf:user() of
         undefined -> wf:redirect("login.htm"), redirect_wait();
         _ -> #dtl{file = "index", app=review,
                   bindings=[{body,body()},
                             {list,list()},
                             {javascript,wf:render(
                                         lists:flatten(
                                         (?MODULE:(wf:config(n2o,mode,dev)))()))}]} end.

prod() ->   [ #script{src="/static/review.min.js"} ].
dev()  -> [ [ #script{src=lists:concat(["/n2o/protocols/",X,".js"])} || X <- [bert,nitrogen] ],
            [ #script{src=lists:concat(["/n2o/",Y,".js"])}           || Y <- [bullet,n2o,ftp,utf8,validation] ] ].

redirect_wait() -> #dtl{}.
list() -> "<iframe src=http://synrc.space/apps/"++code()++" frameborder=0 width=700 height=1250></iframe>".
code() -> case wf:q(<<"code">>) of undefined  -> "../privacy.htm";
                                    Code -> wf:to_list(wf:depickle(Code)) end.

body() ->
    wf:update(heading,#b{id=heading,body="Review: " ++ code()}),
    wf:update(logout,#button{id=logout, body="Logout " ++ wf:user(), postback=logout}),
    [ #span{id=upload},#button { id=send, body= <<"Chat">>, postback=chat, source=[message] } ].

event(init) ->
    Room = code(),
    wf:update(upload,#upload{id=upload}),
    wf:reg(n2o_session:session_id()),
    wf:reg({topic,Room}),
    Res = wf:async("looper",fun index:loop/1),
    n2o_async:send("looper","waterline"),
    wf:info(?MODULE,"Async Process Created: ~p at Page Pid ~p~n",[Res,self()]),
    [ event({client,{E#entry.from,E#entry.media}}) || E <- kvs:entries(kvs:get(feed,{room,Room}),entry,10) ];

event(logout) ->
    wf:logout(),
    wf:redirect("login.htm");

event(chat) ->
    User = wf:user(),
    Message = wf:q(message),
    wf:info(?MODULE,"Chat pressed: ~p~n",[Message]),
    Room = code(),
    kvs:add(#entry{id=kvs:next_id("entry",1),from=wf:user(),feed_id={room,Room},media=Message}),
    wf:send({topic,Room},#client{data={User,Message}});

event(#client{data={User,Message}}) ->
    wf:wire(#jq{target=message,method=[focus,select]}),
    HTML = wf:to_list(Message),
    wf:info(?MODULE,"HTML: ~tp~n",[HTML]),
    DTL = #dtl{file="message",app=review,bindings=[{user,User},{color,"gray"},{message,HTML}]},
    wf:insert_top(history, wf:jse(wf:render(DTL)));

event(#bin{data=Data}) ->
    wf:info(?MODULE,"Binary Delivered ~p~n",[Data]),
    #bin{data = "SERVER"};

event(#ftp{sid=Sid,filename=Filename,status={event,stop}}=Data) ->
    wf:info(?MODULE,"FTP Delivered ~p~n",[Data]),
    Name = hd(lists:reverse(string:tokens(wf:to_list(Filename),"/"))),
    erlang:put(message,wf:render(#link{href=iolist_to_binary(["/static/",Sid,"/",wf:url_encode(Name)]),body=Name})),
    wf:info(?MODULE,"Message ~p~n",[wf:q(message)]),
    event(chat);

event(Event) ->
    wf:info(?MODULE,"Event: ~p", [Event]),
    ok.

loop(M) ->
    DTL = #dtl{file="message",app=review,bindings=[{user,"system"},{message,M},{color,"silver"}]},
    wf:insert_top(history, wf:jse(wf:render(DTL))),
    wf:flush().
